### PR TITLE
Send election result 

### PR DESCRIPTION
**Pull request by celinecamacho**
- Send election result after reception of an election end message. 
- Add and test verify functions for processing of election messages
- Add bad examples of election data messages

### CODE REVIEW

File: be1-go/channel/election/mod.go [hunk 1]
```diff
@@ -304,141 +310,149 @@ func (c *Channel) VerifyPublishMessage(publish method.Publish) error {
...
+		c.log.Error().Msgf("problem decoding sender public key: %v", err)
+		return xerrors.Errorf("sender is %s, should be base64", sender)
 	}
 
-	// note that CreatedAt is provided by the client and can't be fully trusted.
-	// We leave this check as is until we have a better solution.
-	if voteMsg.CreatedAt > c.end {
-		return voteMsg, answer.NewErrorf(-4, "vote cast too late, vote casted at %v "+
-			"and election ended at %v", voteMsg.CreatedAt, c.end)
+	// Check if the sender of the roll call message is the organizer
```
Comment (rjaccard) [inline be1-go/channel/election/mod.go:364]: There is a duplicate comment (lignes 364 and 372). (Here it is not a roll call)

**Review by rjaccard — DISMISSED** [review]
Very cool and very clean PR! And very well tested ! :)

File: be1-go/channel/election/mod.go (hunk 1 above)
Comment (celinecamacho) [reply be1-go/channel/election/mod.go]: Done, thank you

**Review by rjaccard — DISMISSED** [review]

File: be1-go/channel/election/mod.go [hunk 2]
```diff
@@ -173,33 +179,33 @@ func (c *Channel) Publish(publish method.Publish, socket socket.Socket) error {
 	object, action, err := messagedata.GetObjectAndAction(jsonData)
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:179]: Not your fault, but there is a missing error check there that should be added.

File: be1-go/channel/election/mod.go [hunk 3]
```diff
@@ -173,33 +179,33 @@ func (c *Channel) Publish(publish method.Publish, socket socket.Socket) error {
...
-			}
-		case messagedata.ElectionActionResult:
-			err = c.publishResultElection(msg)
-			if err != nil {
-				return xerrors.Errorf("failed to end election: %v", err)
-			}
-		default:
-			return answer.NewInvalidActionError(action)
-		}
+		err = c.processElectionObject(action, msg, socket)
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:182]: I would check the error right bellow instead of outside the if.

File: be1-go/channel/election/mod.go [hunk 4]
```diff
@@ -304,141 +310,149 @@ func (c *Channel) VerifyPublishMessage(publish method.Publish) error {
...
-	for _, q := range voteMsg.Votes {
+	senderPoint := crypto.Suite.Point()
+	err = senderPoint.UnmarshalBinary(senderBuf)
+	if err != nil {
+		return answer.NewError(-4, "invalid sender public key")
+	}
 
-		qs, ok := c.questions[q.Question]
+	// verify sender is an attendee or the organizer
+	ok := c.attendees.IsPresent(msg.Sender) || c.hub.GetPubKeyOrg().Equal(senderPoint)
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:327]: that's elegant

File: be1-go/channel/election/mod.go [hunk 5]
```diff
@@ -304,141 +310,149 @@ func (c *Channel) VerifyPublishMessage(publish method.Publish) error {
...
-func (c *Channel) getAndVerifyCastVoteMessage(msg message.Message) (messagedata.VoteCastVote, error) {
-	var voteMsg messagedata.VoteCastVote
+func (c *Channel) processElectionEnd(msg message.Message) error {
 
-	err := msg.UnmarshalData(&voteMsg)
+	sender := msg.Sender
+	senderBuf, err := base64.URLEncoding.DecodeString(sender)
 	if err != nil {
-		return voteMsg, xerrors.Errorf("failed to unmarshal cast vote: %v", err)
+		c.log.Error().Msgf("problem decoding sender public key: %v", err)
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:360]: is there a reason why we log this specific error and not always the others ?

File: be1-go/channel/election/mod.go [hunk 6]
```diff
@@ -304,141 +310,149 @@ func (c *Channel) VerifyPublishMessage(publish method.Publish) error {
...
 		Questions: []messagedata.ElectionResultQuestion{},
 	}
 
-	for id := range c.questions {
-		question, ok := c.questions[id]
+	for id := range questions {
+		question, ok := questions[id]
 		if !ok {
-			return xerrors.Errorf("No question with this questionId '%s' was recorded", id)
+			return resultElection, xerrors.Errorf("No question with this questionId '%s' was recorded", id)
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:455]: errors must never start with a capital letter

File: be1-go/channel/election/mod.go [hunk 7]
```diff
@@ -500,13 +503,39 @@ func checkMethodProperties(method string, length int) error {
 	return nil
 }
 
-func changeVote(qs *question, earlierVote validVote, sender string, created int64, indexes []int) {
-	if earlierVote.voteTime > created {
-		qs.validVotes[sender] = validVote{
-			voteTime: created,
-			indexes:  indexes,
+func updateVote(msgID string, sender string, castVote messagedata.VoteCastVote, questions *map[string]*question) error {
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:506]: maps are reference types and always passed by reference. So you shouldn't need to pass a pointer of map.

File: be1-go/channel/election/mod.go [hunk 8]
```diff
@@ -500,13 +503,39 @@ func checkMethodProperties(method string, length int) error {
...
+		}
+
+		// this is to handle the case when the organizer must handle multiple
+		// votes being cast at the same time
+		qs.validVotesMu.Lock()
+		earlierVote, ok := qs.validVotes[sender]
+
+		// if the sender didn't previously cast a vote or if the vote is no
+		// longer valid update it
+		if err := checkMethodProperties(qs.method, len(vote.Vote)); err != nil {
```
Comment (nkcr) [inline be1-go/channel/election/mod.go:522]: This is opiniated but we neved use assignment in if statments. That makes the code harder to read.

File: be1-go/channel/election/mod_test.go [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+	"popstellar/message/query/method/message"
+	"popstellar/network/socket"
+	"popstellar/validation"
+	"sync"
+	"testing"
+)
+
+const relativeExamplePath string = "../../../protocol/examples/messageData"
+
+func Test_CastVoteAndResult(t *testing.T) {
```
Comment (nkcr) [inline be1-go/channel/election/mod_test.go:27]: As I told the others it is super helpful to have a small comment about tests to explain what they are testing and doing.

File: be1-go/channel/election/verification_test.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+			err = json.Unmarshal(buf, &castVote)
+			require.NoError(t, err)
+
+			err = electChannel.verifyMessageCastVote(castVote)
+			require.Error(t, err)
+		}
+	}
+
+	t.Run("lao id not base64", getTestBadExample("bad_vote_cast_vote_lao_not_base64.json"))
```
Comment (nkcr) [inline be1-go/channel/election/verification_test.go:58]: nice!

File: be1-go/channel/lao/verification.go [hunk 11]
```diff
@@ -202,3 +208,113 @@ func (c *Channel) verifyMessageRollCallClose(rollCallClose messagedata.RollCallC
...
+	// verify election id is base64URL encoded
+	if _, err := base64.URLEncoding.DecodeString(electionSetup.ID); err != nil {
+		return xerrors.Errorf("election id is %s, should be base64URL encoded", electionSetup.ID)
+	}
+
+	// verify election setup message id
+	expectedID := messagedata.Hash(
+		electionFlag,
+		laoId,
+		fmt.Sprintf("%d", electionSetup.CreatedAt),
```
Comment (nkcr) [inline be1-go/channel/lao/verification.go:236]: For future work, we should `strconv.Itoa()` instead of `fmt.Sprintf()`.

File: be1-go/Makefile [hunk 12]
```diff
@@ -9,7 +9,7 @@ lint:
 	staticcheck ./...
 
 check-cov: protocol lint vet
-	go test -v ./...
+	go test -v -race ./...
```
Comment (nkcr) [inline be1-go/Makefile:12]: I don't understand this makefile. There should be a 

```
check: test test-cov lint vet

test: protocol
	go test -v -race ./...

test-cov: protocol
	go test -v -coverpkg=./... -coverprofile=coverage.out ./... -json > report.json
```

And the CI would only do `make check`. Then we can remove the actual `check` and `check-cov`.

**Review by nkcr — COMMENTED** [review]
Nice work 👍 . You did a great job in adding lots of examples and strengthening the verification process.

File: be1-go/channel/election/verification.go [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	return nil
+}
+
+// verifyRegisteredVotes checks the registered votes of an election end message are valid.
+func verifyRegisteredVotes(electionEnd messagedata.ElectionEnd, questions *map[string]*question) error {
+	// get hashed ID of valid votes sorted by msg ID
+	validVotes := make(map[string]string)
+	validVotesSorted := make([]string, 0)
```
[the remark this answers is not in the document]
Comment (celinecamacho) [reply be1-go/channel/election/verification.go:120]: I don't understand, sorry

File: be1-go/channel/election/mod.go (hunk 5 above)
Comment (celinecamacho) [reply be1-go/channel/election/mod.go:360]: It was to debug at a moment, I can add logs to every error if it's best

File: be1-go/channel/election/verification.go [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+	// verify registered votes are base64URL encoded
+	if _, err := base64.URLEncoding.DecodeString(electionEnd.RegisteredVotes); err != nil {
+		return xerrors.Errorf("election registered votes is %s, should be base64URL encoded", electionEnd.RegisteredVotes)
+	}
+
+	// verify order of registered votes
+	if len(electionEnd.RegisteredVotes) != 0 {
+		err := verifyRegisteredVotes(electionEnd, &c.questions)
+		c.log.Err(err).Msgf("problem with registered votes: %v", err)
```
[the remark this answers is not in the document]
Comment (celinecamacho) [reply be1-go/channel/election/verification.go]: It is not very clear if the votes should be sorted by message ID or timestamp. The e-voting document says timestamp, but a comment from June 8th says message ID. I decided to go with the comment but not be so strong with the checking.

**Review by rjaccard — COMMENTED** [review]
All the utility functions for a list of attendees at the beginning of election/mod.go can be private (or put somewhere else to avoid code duplication?)

**Review by Anghul — COMMENTED** [review]
Nice job, nothing to add on top of what is already said

File: be1-go/channel/election/mod.go (hunk 5 above)
Comment (nkcr) [reply be1-go/channel/election/mod.go:362]: There should be a log only once at the top level, if any.

**Review by nkcr — APPROVED** [review]

**Pull request merged**